### PR TITLE
server: add USR1 signal handler to dump goroutine

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

relate to #7558, we add USR1 signal handler to dump goroutine stack into stdout.

### What is changed and how it works?

- doesn't exit signal handler goroutine until cleanup down
- handle USR1 signal and dump groutine into stdout

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

1) `pkill -USR1 tidb-server` to send signal and check stdout see

```
2018/09/03 11:20:06.150  [info] === Got signal [user defined signal 1] to thread dump===
*** goroutine dump...
goroutine 107 [running]:
main.setupSignalHandler.func1(0xc4200f8e40, 0x14cdd80, 0xc4200dfec0)
	/home/robi/Code/go/src/github.com/pingcap/tidb/tidb-server/main.go:453 +0x2d8
created by main.setupSignalHandler
	/home/robi/Code/go/src/github.com/pingcap/tidb/tidb-server/main.go:446 +0x12e

goroutine 1 [sleep, 10 minutes]:
time.Sleep(0x1176592e000)
	/home/robi/go1.10/src/runtime/time.go:102 +0x166
main.cleanup()
	/home/robi/Code/go/src/github.com/pingcap/tidb/tidb-server/main.go:526 +0x3b
main.main()
	/home/robi/Code/go/src/github.com/pingcap/tidb/tidb-server/main.go:147 +0xbc

goroutine 18 [syscall]:
os/signal.signal_recv(0x14c2840)
	/home/robi/go1.10/src/runtime/sigqueue.go:139 +0xa6
os/signal.loop()
	/home/robi/go1.10/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.0
	/home/robi/go1.10/src/os/signal/signal_unix.go:28 +0x41

goroutine 114 [chan receive]:
github.com/pingcap/tidb/util/systimemon.StartMonitor(0x1410138, 0x140f6a8, 0xc420038f20)
	/home/robi/Code/go/src/github.com/pingcap/tidb/util/systimemon/systime_mon.go:30 +0x154
created by main.setupMetrics
	/home/robi/Code/go/src/github.com/pingcap/tidb/tidb-server/main.go:493 +0xb5
......
``` 

2) temporal add `Sleep(time.Mintue*20) in cleanup(main.go).` 
3) `pkill -QUIT tidb-server` to shutdown tidb-server
4) `pkill -USR1 tidb-server` again and got dump output too

Code changes

 - signal handler change

Side effects

 - no

Related changes

 - no

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/7587)

### CODE REVIEW

File: tidb-server/main.go [hunk 1]
```diff
@@ -432,25 +434,44 @@ func createServer() {
...
-		}
-
-		if xsvr != nil {
-			xsvr.Close() // Should close xserver before server.
+		quited := false
+		for {
+			select {
+			case sig := <-sc:
+				if sig == syscall.SIGUSR1 {
+					buf := make([]byte, 1<<20)
```
Comment (jackysp) [inline tidb-server/main.go]: This may cause a deadlock.

Comment (lysu) [reply tidb-server/main.go]: `make([]byte, 1<<20)` ??

File: tidb-server/main.go [hunk 2]
```diff
@@ -433,20 +432,32 @@ func createServer() {
...
 func setupSignalHandler() {
-	sc := make(chan os.Signal, 1)
-	signal.Notify(sc,
+	dumpSignalChan := make(chan os.Signal, 1)
+	signal.Notify(dumpSignalChan, syscall.SIGUSR1)
+	go func() {
+		buf := make([]byte, 1<<16)
+		for {
+			sig := <-dumpSignalChan
+			if sig == syscall.SIGUSR1 {
```
Comment (coocood) [inline tidb-server/main.go]: We don't need to check the signal.

File: tidb-server/main.go [hunk 3]
```diff
@@ -433,20 +432,32 @@ func createServer() {
...
+				stackLen := runtime.Stack(buf, true)
+				log.Printf("=== Got signal [%s] to goroutine dump===\n*** goroutine dump...\n%s\n*** end\n", sig, buf[:stackLen])
+				continue
```
Comment (coocood) [inline tidb-server/main.go]: `continue` can be removed

File: tidb-server/main.go (hunk 2 above)
Comment (lysu) [reply tidb-server/main.go]: maybe future we need other signal?

File: tidb-server/main.go [hunk 4]
```diff
@@ -48,7 +48,7 @@ import (
 	"github.com/pingcap/tidb/util/printer"
 	"github.com/pingcap/tidb/util/systimemon"
 	"github.com/pingcap/tidb/x-server"
-	binlog "github.com/pingcap/tipb/go-binlog"
+	"github.com/pingcap/tipb/go-binlog"
```
Comment (jackysp) [inline tidb-server/main.go]: This useless alias is used for some contributors to read the code. You could check its commit log.
Btw, I want to remove it either.

File: tidb-server/main.go (hunk 2 above)
Comment (lysu) [reply tidb-server/main.go]: this if and sig local variable is in popose

File: tidb-server/main.go (hunk 4 above)
Comment (lysu) [reply tidb-server/main.go]: .....I didn't modify it...it seem idea or make fmt change it :rofl:  I will rollback it.

**Comment by coocood** [comment]
LGTM

**Comment by lysu** [comment]
@coocood manual test seems be pass too

File: tidb-server/main.go [hunk 5]
```diff
@@ -433,20 +432,31 @@ func createServer() {
...
-	signal.Notify(sc,
+	dumpSignalChan := make(chan os.Signal, 1)
+	signal.Notify(dumpSignalChan, syscall.SIGUSR1)
+	go func() {
+		buf := make([]byte, 1<<16)
+		for {
+			sig := <-dumpSignalChan
+			if sig == syscall.SIGUSR1 {
+				stackLen := runtime.Stack(buf, true)
+				log.Printf("=== Got signal [%s] to goroutine dump===\n*** goroutine dump...\n%s\n*** end\n", sig, buf[:stackLen])
```
Comment (zz-jason) [inline tidb-server/main.go]: how about:
```
=== Got signal [%s] to dump goroutine stack. ===
%s
=== Finished dumping goroutine stack. ===
```

Comment (lysu) [reply tidb-server/main.go]: Done~~

**Pull request merged**